File: bash-verifier/lib/f32/f32.c
#include <stdlib.h>

#include <caml/alloc.h>
#include <caml/memory.h>
#include <caml/mlvalues.h>

#define fabs_f(x) (x < 0.0f ? - x : x)

union int_float { float f; int32_t i; };

static inline float float_of_int(int32_t i) {
  union int_float x = { .i = i };
  return x.f;
}

static inline int32_t int_of_float(float f) {
  union int_float x = { .f = f };
  return x.i;
}

#define F32_val(v)  (float_of_int(Int32_val(v)))
#define F32_copy(f) (caml_copy_int32(int_of_float(f)))

CAMLprim value f32_add(value x, value y) {
  CAMLparam2(x, y);
  CAMLreturn (F32_copy (F32_val(x) + F32_val(y)));
}

CAMLprim value f32_sub(value x, value y) {
  CAMLparam2(x, y);
  CAMLreturn (F32_copy (F32_val(x) - F32_val(y)));
}

CAMLprim value f32_mul(value x, value y) {
  CAMLparam2(x, y);
  CAMLreturn (F32_copy (F32_val(x) * F32_val(y)));
}

CAMLprim value f32_div(value x, value y) {
  CAMLparam2(x, y);
  CAMLreturn (F32_copy (F32_val(x) / F32_val(y)));
}

CAMLprim value f32_abs(value x) {
  CAMLparam1(x);
  CAMLreturn (F32_copy (fabs_f(F32_val(x))));
}

CAMLprim value f32_neg(value x) {
  CAMLparam1(x);
  CAMLreturn (F32_copy (- F32_val(x)));
}

CAMLprim value f32_of_double(value x) {
  CAMLparam1(x);
  CAMLreturn (F32_copy((float) Double_val(x)));
}

CAMLprim value f32_to_double(value x, value y) {
  CAMLparam2(x, y);
  CAMLreturn (caml_copy_double((double) F32_val(x)));
}

File: bash-verifier/lib/f32/f32.ml
type t = int32

external add : t -> t -> t = "f32_add"
external sub : t -> t -> t = "f32_sub"
external mul : t -> t -> t = "f32_mul"
external div : t -> t -> t = "f32_div"

external abs : t -> t = "f32_abs"
external neg : t -> t = "f32_neg"

external of_float : float -> t = "f32_of_double"
external to_float : t -> float = "f32_to_double"

let compare : t -> t -> int = Stdlib.compare

let zero = of_float 0.0
let one  = of_float 1.0

let ( + ) = add
let ( - ) = sub
let ( * ) = mul
let ( / ) = div

let of_string s = of_float (Float.of_string s)
let of_string_opt s = Option.map of_float (Float.of_string_opt s)
let to_string f = Float.to_string (to_float f)
let printer c f = Printf.fprintf c "%f" (to_float f)
